add drag estimate scalar

fredmarkus opened the pull request:
Adds a drag scalar that is estimated as part of exponential drag model for the EKF.
I tried using J2 but it significantly increased run duration and did not improve performance much at all. Not worth incorporating for now. J3 and J4 also seem to have no numerical effect (they evaluate to 0). This may improve when using natural units.

- [x] Add third body gravitational accelerations (Sun, Moon)
- [x] Include J3, J4 if performance doesn't suffer too much 
- [x] Add explanation and derivation for drag model

Solves #115, #116 

Explanation of exponential model and partial derivatives

![image](https://github.com/user-attachments/assets/fa5acd70-72bc-4371-ac27-5a28c4de7b33)

![image](https://github.com/user-attachments/assets/05dd1555-d496-441a-9c7d-3e2431b6dfaf)

![image](https://github.com/user-attachments/assets/9797cbe6-398a-417a-9ef2-e452fb06f591)

fredmarkus commented:
@amaarquadri It seems that the ground truth drag model term computed in `drag_dynamics` is smaller than the J2 term. I'm not sure that is correct, but it maybe considering the relatively small surface area of our spacecraft.

amaarquadri commented:
> @amaarquadri It seems that the ground truth drag model term computed in `drag_dynamics.drag_dynamics` is smaller than the J2 term. I'm not sure that is correct, but it maybe considering the relatively small surface area of our spacecraft. The cause of this would lie in the atmospheric density parameter. Please take a look if you can see anything wrong with the Harries-Priester model.

I found an index out of bound error on [this line](https://github.com/cmu-argus-2/GNC-Payload/blob/main/utils/earth_utils.py#L481), but its only relevant near the very top of the atmospheric model so its not the root cause.

I would compare against the implementation in GNC-Simulation [here](https://github.com/cmu-argus-2/GNC-Simulation/blob/main/argusim/world/physics/models/drag.cpp#L24). Or even better, since we don't need to be able to differentiate the density model (since we'll be using a simpler exponential model in the filter) we can just use their code as a dependency. They already have it set up as a package too.

amaarquadri commented on dynamics/third_body_dynamics.py:
```diff
@@ -0,0 +1,61 @@
+
+import brahe
+import numpy as np
+from brahe import GM_MOON, GM_SUN, Epoch
```
```suggestion
from brahe import GM_MOON, GM_SUN
from brahe.epoch import Epoch
```

amaarquadri commented on dynamics/test_j2_dynamics.py:
Can you please rename this to `test_grav_potential_dynamics.py`?

amaarquadri commented on dynamics/orbital_dynamics.py:
```diff
@@ -189,6 +212,28 @@ def perturbed_state_derivative(self, x: np.ndarray, epoch: Epoch = None) -> np.n
+        if self.use_j34 and not np.isclose(r_norm, 0):
+            a_J3_gt = j3_dynamics(r)
+            a_J4_gt = j4_dynamics(r)
+            print(f"J3: {a_J3_gt}, J4: {a_J4_gt}")
```
```suggestion
```

amaarquadri commented on dynamics/orbital_dynamics.py:
```diff
@@ -189,6 +212,28 @@ def perturbed_state_derivative(self, x: np.ndarray, epoch: Epoch = None) -> np.n
+            updated_a += a_J3_gt + a_J4_gt
+
+        # Compute third body gravity
+        if self.use_sun_grav and not np.isclose(r_norm, 0):
```
I don't think you need this extra check since there shouldn't be any division by zero issues here.
```suggestion
        if self.use_sun_grav:
```

amaarquadri commented on dynamics/orbital_dynamics.py:
```diff
@@ -189,6 +212,28 @@ def perturbed_state_derivative(self, x: np.ndarray, epoch: Epoch = None) -> np.n
+
+            updated_a += a_sun_gt
+
+        if self.use_moon_grav and not np.isclose(r_norm, 0):
```
```suggestion
        if self.use_moon_grav:
```

amaarquadri commented on dynamics/orbital_dynamics.py:
```diff
@@ -224,6 +269,28 @@ def perturbed_state_derivative_jac(self, x: np.ndarray, epoch: Epoch = None) ->
+            da_dr += da_J3_gt_dr + da_J4_gt_dr
+
+        # Compute third body gravity
+        if self.use_sun_grav and not np.isclose(np.linalg.norm(x[0:3]), 0):
```
```suggestion
        if self.use_sun_grav:
```

amaarquadri commented on dynamics/orbital_dynamics.py:
```diff
@@ -224,6 +269,28 @@ def perturbed_state_derivative_jac(self, x: np.ndarray, epoch: Epoch = None) ->
+
+            da_dr += da_sun_gt_dr
+
+        if self.use_moon_grav and not np.isclose(np.linalg.norm(x[0:3]), 0):
```
```suggestion
        if self.use_moon_grav:
```

amaarquadri commented on dynamics/orbital_dynamics.py:
```diff
@@ -239,7 +306,7 @@ def perturbed_f(self, x: np.ndarray, dt: float, epoch: Epoch = None) -> np.ndarr
         func = (
             partial(self.perturbed_state_derivative, epoch=epoch)
-            if self.use_drag
+            if (self.use_drag or self.use_moon_grav or self.use_sun_grav)
```
I think it might make sense to introduce a `requires_epoch` member function for this to avoid code duplication.

fredmarkus replied:
You mean duplication of the condition?

amaarquadri replied:
Yes.

amaarquadri commented on dynamics/grav_potential_dynamics.py:
```diff
@@ -30,13 +94,6 @@ def j2_dynamics(r: jnp.ndarray) -> jnp.ndarray:
 
 
 def j2_jacobian_auto(r: jnp.ndarray) -> jnp.ndarray:
-    """
```
Was this intentional?

fredmarkus replied:
Yes. I removed the docstrings for functions of 3 lines or less in this file. Also changed the pylintrc to reflect this.

amaarquadri replied:
I'm not sure I'm in favor of this. Docstrings only take a few seconds to add, especially with AI autocomplete nowadays.

amaarquadri commented on dynamics/third_body_dynamics.py:
```diff
@@ -0,0 +1,61 @@
+    """
+    Computes the third body acceleration.
+
+    :param x: state vector
```
Can you rename this to `r_sat` or something? There's no need for knowledge of the full state vector here. Same goes for the rest of the functions in this file.

amaarquadri commented on dynamics/drag_dynamics.py:
```diff
@@ -52,3 +55,70 @@ def drag_jacobian(x: np.ndarray, drag_const: float, latest_epoch: Epoch) -> np.n
     return da_drag_dv
+
+
+def drag_scalar_estimate(x: np.ndarray, drag_const: float) -> np.ndarray:
```
Can you change all the inputs here to be specifically what the functions require, not just the state vector? These functions are low level building blocks and should not have any knowledge of higher level concepts like the state vector.

I can't even tell what `x[9]` is in this function without having to dig through the EKF code, which I think illustrates why separation of concerns like this is essential. Same thing goes for the functions that follow in this file.

amaarquadri reviewed (changes requested):
I only reviewed the files with comments. Once these adjustments are in, it'll be easier for me to understand the code and review the rest.

fredmarkus commented:
@amaarquadri I have added a comment regarding the drag_jacobian function. As the harris-priester density function uses position and velocity as arguments, we technically need to take it into account in the drag jacobian. However, the function itself is not smooth so we can differentiate it. 
however, since the function is not used anyways at the moment (since the groundtruth dynamics doesn't require jacobians), i have just added a comment.

fredmarkus commented:
@amaarquadri all resolved

amaarquadri commented on .pylintrc:
```diff
@@ -180,7 +180,7 @@ const-naming-style=UPPER_CASE
 # Minimum line length for functions/classes that require docstrings, shorter
 # ones are exempt.
-docstring-min-length=-1
+docstring-min-length=3
```
```suggestion
docstring-min-length=-1
```

amaarquadri commented on dynamics/orbital_dynamics.py:
```diff
@@ -31,24 +45,38 @@ def __init__(
             / config["satellite"]["mass"]
         )
 
+        if self.use_drag or self.use_sun_grav or self.use_moon_grav:
+            self.require_epoch = True
+        else:
+            self.require_epoch = False
+
```
```suggestion
    @property
    def require_epoch(self) -> bool:
        """
        :return: True if the configured perturbations require the current time epoch, False otherwise.
        """
        return self.use_drag or self.use_sun_grav or self.use_moon_grav
```

fredmarkus replied:
clean

amaarquadri commented on dynamics/drag_dynamics.py:
```diff
@@ -52,3 +60,79 @@ def drag_jacobian(x: np.ndarray, drag_const: float, latest_epoch: Epoch) -> np.n
+
+    v_norm = np.linalg.norm(x[3:6])
+    height = np.linalg.norm(x[0:3]) - R_EARTH
+    density_estimate = d_est * NOMINAL_DENSITY * np.exp(-height / REF_HEIGHT)
```
If I'm reading your latex images right, shouldn't this be `np.exp(-(height - REF_HEIGHT) / some_constant)`. As is, you'll get an extra factor of 1/e when `height == REF_HEIGHT`.

fredmarkus replied:
No this formulation is correct. R_E refers to the earth radius R_EARTH and H is the Reference Height REF_HEIGHT.
so essentially if you're right on the earth's surface, the exponential term becomes 1 and the further away you go from the surface, the smaller the density term becomes

amaarquadri replied:
Isn't the expectation that at REF_HEIGHT the density is NOMINAL_DENSITY (if d_est=1)? Currently, you get an extra factor of 1/e.

fredmarkus replied:
d_est is not linked to the nominal density. I took the formulation from Montenbuck and Gill. I can check the book again tomorrow in the lab.

amaarquadri reviewed (changes requested):
Potential bug in the exponential drag model (with downstream impacts on the corresponding derivatives). Other than that and a few nitpicks, lgtm.

fredmarkus commented:
@amaarquadri Added a description on the drag estimate function for more clarity. Hope this clears it up.

amaarquadri reviewed (approved):
LGTM, great work!
